libs/tlmatcher/src/main/java/org/codice/util/tlmatcher/TimeLimitedMatcher.java

```diff
@@ -0,0 +1,229 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.util.tlmatcher;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Customized and reduced-scope RegEx matcher class to protect against runaway searches.
+ *
+ * <p>Instances of this class, created through one of its two factory methods, will either succeed
+ * or timeout, throwing a RuntimeException - specifically {@link TimeoutException} - in the latter
```

Comment:
✏️ There's a subtle difference here from the `Matcher` class that I think should be called out in the javadoc: the fact that the factory methods actually check for matching rather than the `matches` method (and throw the `TimeoutException`). It may also make that more clear if you put the `if (matcher.matches())` check outside in the try-catch block in your example.

Reply:
Good suggestion. 👍

Reply:
Yeah! My comment was about moving the actual matching where users would be expecting it to be done ... in the `match()` method as opposed to the factory method.

Reply:
Why not throw a checked exception since we are wanting to ensure callers handle that case?

Reply:
I have an aversion to checked exceptions. If they don't handle it, it still bubbles up; handling it is just more graceful.

Reply:
But they have to explicitly handle and re throw a checked exception to have it bubble up. as opposed to not having to do anything.

Reply:
I tend to agree with @rzwiefel here, a checked exception would be better than a runtime one since we explicitly want them to handle it or do something about it. A runtime exception will be left ignored which is not necessarily a good thing to do.

Reply:
That's actually their choice. The RTException _will_ propagate all the way out if they don't catch it.

Reply:
That is what is bad about it. Something at the top shouldn't be aware that someone down below was doing pattern matching that took too long. That is an implementation detail.

Reply:
Y'all feel more strongly about this than I do, so I can change it.